File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
       # traversal algorithm here is: bredth first 
       # we want to change the swap the right and left childs
       # O(n) => Time
       # O(n) => Space
       # solve it recursively DFS

       # base condition
       if root is None: return None

       tmp: Optional[TreeNode] = root.left
       root.left = root.right
       root.right = tmp

       self.invertTree(root.left)
       self.invertTree(root.right)

       return root

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */


class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0 || inorder.length == 0) return null;

        TreeNode root = new TreeNode(preorder[0]);

        int midIdx = 0;
        for (int i = 0; i < inorder.length; i++) {
            if (inorder[i] == root.val) {
                midIdx = i;
                break;
            }
        }

        int[] leftPre = Arrays.copyOfRange(preorder, 1, midIdx + 1);
        int[] leftIn = Arrays.copyOfRange(inorder, 0, midIdx);

        int[] rightPre = Arrays.copyOfRange(preorder, midIdx + 1, preorder.length);
        int[] rightIn = Arrays.copyOfRange(inorder, midIdx + 1, inorder.length);

        root.left = buildTree(leftPre, leftIn);
        root.right = buildTree(rightPre, rightIn);

        return root;
    }
}